### ISSUE TITLE
How to properly disconnect and reconnect to the Stream Deck App?

### ISSUE THREAD

**Issue by tinodo** [TyrenDe/streamdeck-client-csharp#14]
I am using a plugin to connect to an external application. The external application uses the streamdeck-client-sharp library to connect to the Stream Deck App. That application might be started, stopped and restarted. The plugin sends the connection information received by the Stream Deck App to the external application.
I want to be able to properly disconnect from the Stream Deck App, and to be able to reconnect again.

When I use the StreamDeckConnection.Close(), where the m_CancelSource.Cancel() is called, and try to connect again using a NEW StreamDeckConnection and StreamDeckConnection.Run(), I see this in the Stream Deck logs:

ESDCustomPluginCommunication::onMessageReceived(): Plugin 703EA06C6B1F93B5DC9A996984D6AEFF is already connected

And of course, the new connection doesn't work as expected.
Most likely, this is caused by this line in the RunAsync() method:
await SendAsync(new RegisterEventMessage(m_RegisterEvent, this.UUID));
It sends the RegisterEventMessage which I think triggers the error in the Stream Deck App side.
Would it be enough to simply reconnect the websocket connect on the proper port? Without 'registering'?

What is the proper way to disconnect the StreamDeckConnection from the Stream Deck App to be able to reconnect?

Thanks!

**Comment by tinodo** [comment]
Okay, so apparently there is no proper way to reconnect, other than to just crash the plugin. So that is what I implemented when I see that "Plugin 703...EFF is already connected" message. Just crash the plugin, the Stream Deck App will restart the plugin and you will get a new context.

**Comment by BarRaider** [comment]
I think you're looking at this the wrong way. 
Your plugin should not be disconnecting from the stream deck nor crashing itself. The stream deck is the one that starts up the plugin and passes the information to the plugin on how to "register". From that point onwards you should remain connected until the stream deck disconnects you.

> That application might be started, stopped and restarted. 

That has nothing to do with the plugin. The application is separate from the plugin. If the plugin cannot connect to the application for whatever reason, it should still stay connected to the stream deck but maybe show an icon/title indicating this. Similarly if the application restarts, the plugin should refresh the data but not by crashing itself and restarting

**Comment by tinodo** [comment]
Hi @BarRaider,

I get the idea, but you cannot re-connect to the Stream Deck App using the same port and UUID twice. So after an initial successful connection, should you get disconnected, you cannot simply re-establish a websocket connection to the Stream Deck App again. If you try, you get the "Plugin is already connected" error. The only way I was able to get a new UUID, is to crash the actual executable that the Stream Deck App monitors. It will then restart that app with a new UUID in the command line arguments.

If you feel this is incorrect, I'd be eager to learn more!

For clarity; my actual plugin (the executable in the %appdata%\Roaming\Elgato\StreamDeck\Plugins\\[something] and mentioned in the manifest.json) parses the startup parameters (which includes a port and a UUID). I tries to send this information to the application that implements the Stream Deck Client (SDK). It will retry that operation until the application responds. The application will connect to the Stream Deck App using the StreamDeckConnection.Run() method. Should that application restart, it cannot reconnect using the same port and UUID... The only way I found to get a new port and UUID is through restarting the Plugin... Again, if that's wrong; I'd love to learn the proper way to do this!

**Comment by BarRaider** [comment]
> should you get disconnected, you cannot simply re-establish a websocket connection

Why would this ever happen?
Here is how I handle it in Stream deck tools
Notice it just gracefully exits but it's because the connection with the stream deck ended, not my integrated application.

https://github.com/BarRaider/streamdeck-tools/blob/7cdb26bb2b6c727a0492465d7b0e3b1465871498/barraider-sdtools/Backend/PluginContainer.cs#L68

**Comment by tinodo** [comment]
The implementation of the Stream Deck SDK is done in a different, completely separate, Windows application. It supports integration with many, many other tools and (enterprise) applications (Yes; THANK YOU for the OBS Studio libraries as well!). So that application can be closed and started at any given point by a user. It is NOT started or stopped alongside the Stream Deck App nor the 'actual' plugin for the Stream Deck App. (Touch Portal is easier; you can just connect, disconnect and reconnect whenever you feel like it. So the TouchPortal Plugin I create is just an entry.tp file, not even an executable. The implementation is in that separate Windows App I wrote. You could consider this application to be close to streamerbot or something like that.

But I feel that my situation is rather a niche thing. I don't want to take too much of your time. Thanks for all the help and explanation. Much appreciated!

**Comment by BarRaider** [comment]
Sure thing. If you have further questions you can join the #developers-chat room on our Discord. http://discord.barraider.com

**Comment by fifafu** [comment]
I have the exactly same situation, I'm currently integrating Stream Deck support in BetterTouchTool (https://folivora.ai), but I'd like to keep everything in the main application to reduce expensive IPC. The plugin just forwards the registration data to the main app. This works well, until the main app is restarted for whatever reason. Then I get the "Plugin is already connected" message and the only way to reconnect seems to crash the plugin.

I would only need a proper disconnect function, then my setup would work really well. Unfortunately it looks like the Stream Deck software does not support this yet, so nothing this library can do about it.

**Comment by tinodo** [comment]
I got this issue resolved by writing a small muck plugin that communicates with the actual application implementing all the plugin logic. It gets the connection information once it connects to the Stream Deck app and stores that information. When the actual application wants to connect, it gets the stored connection information from the plugin and uses that information to connect.

I'm using RPC for that communication, but it's up to you. I have implemented 2 methods in the actual small plugin; start and restart. Restart closes the plugin forcing it to restart and reconnect to Stream Deck hence obtaining new connection information. Start passes the obtained connection information from Stream Deck.

This way, when the actual plugin starts before your application, it sends the obtained connection information to the application through the Start method. If you application restarts after that, and new connection information needs to be obtained, detect the problem in your actual application and send the Restart command to the plugin; it will close and Stream Deck will restart it; it obtains new connection information and will send that to the application.

Hope that process is clear and it helps!
